core/src/cluster_info.rs

```diff
@@ -2272,95 +2276,114 @@ impl ClusterInfo {
+            let now = timestamp();
+            messages
+                .into_iter()
+                .flat_map(|(from, crds_values)| {
+                    gossip.process_push_message(&from, crds_values, now)
+                })
+                .map(|v| v.value.label())
+                .collect()
+        };
+        // Generate prune messages.
+        let prunes_map = self
             .time_gossip_write_lock("prune_received_cache", &self.stats.prune_received_cache)
             .prune_received_cache(updated_labels, stakes);
-
-        let rsp: Vec<_> = prunes_map
-            .into_iter()
-            .filter_map(|(from, prune_set)| {
-                inc_new_counter_debug!("cluster_info-push_message-prunes", prune_set.len());
-                self.lookup_contact_info(&from, |ci| ci.clone()).map(|ci| {
-                    let mut prune_msg = PruneData {
-                        pubkey: self_id,
-                        prunes: prune_set.into_iter().collect(),
-                        signature: Signature::default(),
-                        destination: from,
-                        wallclock: timestamp(),
-                    };
-                    prune_msg.sign(&self.keypair);
-                    let rsp = Protocol::PruneMessage(self_id, prune_msg);
-                    (ci.gossip, rsp)
-                })
+        let prune_messages: Vec<_> = {
+            let gossip = self.gossip.read().unwrap();
+            let wallclock = timestamp();
+            let self_pubkey = self.id();
+            thread_pool.install(|| {
+                Vec::from_iter(prunes_map)
+                    .into_par_iter()
+                    .with_min_len(256)
+                    .filter_map(|(from, prune_set)| {
+                        let peer = gossip.crds.get_contact_info(&from)?;
+                        let mut prune_data = PruneData {
+                            pubkey: self_pubkey,
+                            prunes: Vec::from_iter(prune_set),
+                            signature: Signature::default(),
+                            destination: from,
+                            wallclock,
+                        };
+                        prune_data.sign(&self.keypair);
+                        let prune_message = Protocol::PruneMessage(self_pubkey, prune_data);
+                        Some((peer.gossip, prune_message))
+                    })
+                    .collect()
             })
-            .collect();
-        if rsp.is_empty() {
-            return None;
+        };
+        if prune_messages.is_empty() {
+            return;
         }
-        let mut packets = to_packets_with_destination(recycler.clone(), &rsp);
+        let mut packets = to_packets_with_destination(recycler.clone(), &prune_messages);
         self.stats
             .push_response_count
             .add_relaxed(packets.packets.len() as u64);
-        if !packets.is_empty() {
```

Comment:
just to clarify, this `is_empty` check no longer necessary because the

```if prune_messages.is_empty()```

above guarantees the returned `packets` it's not empty?

Reply:
yes